DataViews: Register the deletePost action like any third-party action.

youknowriad opened the pull request:
Related #61084 
Similar to #62647 

## What?

In #62052 an API to register and unregister dataviews actions has been implemented. But in order to allow third-party developers to be able to unregister these actions, we need to be using the same actions in Core to register the core actions.

The current PR explore the possibility to use the API to register one action: "delete post". 

## Testing Instructions

1- Open the patterns dataviews and try removing a user created pattern.
2- You should be able to see the "delete" action in the actions dropdown menu
3- you can try to use the action.

youknowriad committed:
DataViews: Register the deletePost action like any third-party action.

youknowriad commented:
The other similar PR https://github.com/WordPress/gutenberg/pull/62647 got blocked because we need access to the store and we need isElligible to be reactive (which is hard to solve). 

So I'm trying another action that doesn't have these requirements in this PR. But the current PR is also blocked for two reasons:

 - I'm importing JS files in TS files and would like to ignore the JS files in the typing build but it doesn't seem to work so far.
 - I noticed that the template parts dataviews don't pass the "template part" objects as "items", instead they nest them inside `item.templatePart` which is a no-go for me and I'm trying to understand why we're doing this.

sirreal commented on packages/editor/src/dataviews/types.ts:
```diff
@@ -0,0 +1,27 @@
+
+export interface OtherPost extends BasePost {
+	type: Exclude< string, ReservedPostTypes >;
+}
```
I think we can simplify this section, the following might work:

```suggestion
export interface Post {
	status?: PostStatus;
	title: string | { rendered: string };
	type: string;
}

export interface TemplateOrTemplatePart extends Post {
	type: 'template' | 'template-part';
	source: string;
	has_theme_file: boolean;
}
```

The main reason is that `Exclude< string, ReservedPostTypes >` can't really do anything to narrow the type, it's just `string`. The type "any string except these two literals" is too broad and just becomes `string`. Any `OtherPost` would also satisfy the `TemplateOrTemplatePart` type.

A type guard would be suitable to narrow the type in application code:

```tsx
function isTemplateOrTemplatePartPost( p: Post ): p is TemplateOrTemplatePart {
    return p.type === 'template' || p.type === 'template-part';
}

declare var p: Post;
if ( isTemplateOrTemplatePartPost(p)) {
    p.has_theme_file; // Now `p` is of type `TemplateOrTemplatePart`
}
```

sirreal committed:
Include `src` files but do not check js

sirreal committed:
Patch postcss-urlrebase package to remove node URL reference

The block editor uses this package in the browser.
URL is available globally, there is no need to import it.

sirreal committed:
Remove reference to missing type WPBlock

youknowriad committed:
Changes per review

youknowriad committed:
Simplify registration of actions

youknowriad commented:
This should be ready for review.

youknowriad commented on packages/edit-post/src/index.js:
```diff
@@ -91,6 +93,7 @@ export function initializeEditor(
 			enableFSEBlocks: settings.__unstableEnableFullSiteEditingBlocks,
 		} );
 	}
+	registerDefaultActions();
```
I couldn't find a better place to call this. I kind of maps what we do for blocks.

youknowriad commented on packages/editor/src/dataviews/types.ts:
```diff
@@ -0,0 +1,21 @@
+	has_theme_file: boolean;
+}
+
+export type Post = TemplateOrTemplatePart | BasePost;
```
I like a lot that we could use the "actions" refactoring to slowly augment our "entities" types. In the long run, this will help a lot when build UI for entities.

jorgefilipecosta commented on patches/postcss-urlrebase+1.3.0.patch:
```diff
@@ -0,0 +1,11 @@
+diff --git a/node_modules/postcss-urlrebase/index.d.ts b/node_modules/postcss-urlrebase/index.d.ts
```
Why is this change required?

youknowriad replied:
Something about fixing the typescript types. But @sirreal would know more precisely.

sirreal replied:
A browser package (editor, I believe) uses postcss in the browser.

This dependency expected to work in a node environment (most common for postcss stuff) and use the URL type imported from node. This is not available in the environment. However, it's fine to use the URL global, which is what this change does. Otherwise, the package would error when it tried to import a non-existent (node.js) type.

sirreal replied:
https://github.com/strarsis/postcss-urlrebase/pull/2

sirreal replied:
Add package patch note here: https://github.com/WordPress/gutenberg/pull/63015

jorgefilipecosta reviewed (approved):
Everything worked well on the tests I did 👍

ntsekouras commented on packages/editor/src/dataviews/actions/utils.ts:
```diff
@@ -0,0 +1,50 @@
+			TEMPLATE_ORIGINS.custom
+		) &&
+		! template.has_theme_file &&
+		! template?.has_theme_file
```
This is the same check now.. https://github.com/WordPress/gutenberg/pull/63021
